diff --git a/src/drivers/driver_circle.cpp b/src/drivers/driver_circle.cpp
index 1a02dc4f0..79906c26a 100644
--- a/src/drivers/driver_circle.cpp
+++ b/src/drivers/driver_circle.cpp
@@ -341,7 +341,7 @@ static int wpa_driver_circle_scan2 (void *priv, wpa_driver_scan_params *params)
 	//assert (params->num_ssids == 0);
 	assert (params->extra_ies == 0);
 	assert (params->extra_ies_len == 0);
-	assert (params->freqs == 0);
+	//assert (params->freqs == 0);
 
 	assert (drv->netdev != 0);
 	// increase scan duration here to be sure, scan is not started again
